Generate secure coding serializer for DDActionContext (and the types it relies on)

beidson opened the pull request:
#### 5776a8fdee1ed5230bafbb063ad15ac71d607f55
<pre>
Generate secure coding serializer for DDActionContext (and the types it relies on)
<a href="https://bugs.webkit.org/show_bug.cgi?id=265890">https://bugs.webkit.org/show_bug.cgi?id=265890</a>
<a href="https://rdar.apple.com/119203785">rdar://119203785</a>

Reviewed by Alex Christensen.

This adds the WebKit property list serialization code path for DDActionContext.
That includes the types it relies on - NSPersonNameComponents, and NSRect wrapped by NSValue.

On the surface - by looking at how our API test works - it appears almost as if we could serialize DDActionContext
directly by accessing members and recreating with setters.
Unfortunately that&apos;s not the case because not everything they need to encode/decode is directly accessible.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::getClass&lt;WKDDActionContext&gt;):
(IPC::typeFromObject):
* Source/WebKit/Shared/Cocoa/CoreIPCDDActionContext.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.mm:
(WebKit::CoreIPCDictionary::keyHasValueOfType const):
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
(WebKit::valueFromID):
* Source/WebKit/Shared/Cocoa/CoreIPCNSValue.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSValue.mm:
(WebKit::CoreIPCNSValue::valueFromNSValue):
(WebKit::CoreIPCNSValue::toID const):
(WebKit::CoreIPCNSValue::shouldWrapValue):
* Source/WebKit/Shared/Cocoa/CoreIPCPersonNameComponents.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCNSValue.h.
(WebKit::CoreIPCPersonNameComponents::CoreIPCPersonNameComponents):
* Source/WebKit/Shared/Cocoa/CoreIPCPersonNameComponents.mm: Copied from Source/WebKit/Shared/Cocoa/CoreIPCNSValue.mm.
(WebKit::CoreIPCPersonNameComponents::CoreIPCPersonNameComponents):
(WebKit::CoreIPCPersonNameComponents::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCPersonNameComponents.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCTypes.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(isEqual):
(operator==):
(personNameComponentsForTesting):
(TEST):
(fakeDataDetectorResultForTesting):

Canonical link: <a href="https://commits.webkit.org/271585@main">https://commits.webkit.org/271585@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4873a9daf261bfb6e37706812559a7730b5cc976

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28843 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7485 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30214 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31460 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26310 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29415 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9619 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4844 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26356 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29114 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6202 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24782 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5390 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5539 "Found 5 new test failures: imported/w3c/web-platform-tests/css/css-view-transitions/massive-element-below-viewport-partially-onscreen-new.html, imported/w3c/web-platform-tests/css/css-view-transitions/massive-element-below-viewport-partially-onscreen-old.html, imported/w3c/web-platform-tests/html/semantics/disabled-elements/disabled-event-dispatch.tentative.html, imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-button-event-dispatch.tentative.html, imported/w3c/web-platform-tests/html/semantics/selectors/pseudo-classes/readwrite-readonly.html (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25793 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/32798 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26398 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26235 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31779 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5510 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3678 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29560 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7137 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6901 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5987 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6031 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
```diff
@@ -276,6 +283,10 @@ NSType typeFromObject(id object)
         return NSType::Color;
 #if ENABLE(DATA_DETECTION)
+#if PLATFORM(MAC)
+    if (PAL::isDataDetectorsCoreFrameworkAvailable() && [object isKindOfClass:[PAL::getWKDDActionContextClass() class]])
```
Here and below, I'm surprised that the class selector call is necessary.

beidson replied:
I'm not sure I follow the comment.
The purpose of the function is to match classes to enum values.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCDictionary.mm:
```diff
@@ -55,8 +55,11 @@
-    if (!object)
-        return false;
+    if (!object) {
+        // Many objects have a required key that sometimes has a missing value, which is okay.
```
doesn't that mean it's not a required key?

beidson replied:
It's ill formed for the key to be missing.
But the key can be null.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCPersonNameComponents.mm:
```diff
@@ -0,0 +1,62 @@
+
+RetainPtr<id> CoreIPCPersonNameComponents::toID() const
+{
+    RetainPtr<NSPersonNameComponents> components = adoptNS([[NSPersonNameComponents alloc] init]);
```
auto components = adoptNS([NSPersonNameComponents new]); ?

beidson replied:
Newfangled stylist.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCPersonNameComponents.serialization.in:
```diff
@@ -0,0 +1,36 @@
+    String m_middleName;
+    String m_familyName;
+    String m_nickname;
+    std::unique_ptr<WebKit::CoreIPCPersonNameComponents> m_phoneticRepresentation;
```
Can this be infinitely nested, or just one nested?  We might want to make a struct that contains the names, then make CoreIPCPersonNameComponents one of those and optionally another one of those.

beidson replied:
I think infinitely nested, but am checking.

beidson replied:
I'm leaving it as is with the question to that team outstanding. Can loop back around if it needs to change.

achristensen07 commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -18675,7 +18687,7 @@
 			platformFilters = (
 				ios,
-				xros,
+                xros,
```
This indentation isn't good.

beidson replied:
Correct, it isn't

achristensen07 commented on Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
```diff
@@ -175,6 +178,40 @@ id valueAsID() const
+        if (a) {
+            if (!b)
+                return false;
+            if (!CFEqual(a, b))
```
Doesn't CFEqual already handle the cases where one or both is null?

beidson replied:
You... are probably right lol.

beidson committed:
Generate secure coding serializer for DDActionContext (and the types it relies on)
https://bugs.webkit.org/show_bug.cgi?id=265890
rdar://119203785

Reviewed by Alex Christensen.

This adds the WebKit property list serialization code path for DDActionContext.
That includes the types it relies on - NSPersonNameComponents, and NSRect wrapped by NSValue.

On the surface - by looking at how our API test works - it appears almost as if we could serialize DDActionContext
directly by accessing members and recreating with setters.
Unfortunately that's not the case because not everything they need to encode/decode is directly accessible.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::getClass<WKDDActionContext>):
(IPC::typeFromObject):
* Source/WebKit/Shared/Cocoa/CoreIPCDDActionContext.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.mm:
(WebKit::CoreIPCDictionary::keyHasValueOfType const):
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
(WebKit::valueFromID):
* Source/WebKit/Shared/Cocoa/CoreIPCNSValue.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSValue.mm:
(WebKit::CoreIPCNSValue::valueFromNSValue):
(WebKit::CoreIPCNSValue::toID const):
(WebKit::CoreIPCNSValue::shouldWrapValue):
* Source/WebKit/Shared/Cocoa/CoreIPCPersonNameComponents.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCNSValue.h.
(WebKit::CoreIPCPersonNameComponents::CoreIPCPersonNameComponents):
* Source/WebKit/Shared/Cocoa/CoreIPCPersonNameComponents.mm: Copied from Source/WebKit/Shared/Cocoa/CoreIPCNSValue.mm.
(WebKit::CoreIPCPersonNameComponents::CoreIPCPersonNameComponents):
(WebKit::CoreIPCPersonNameComponents::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCPersonNameComponents.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCTypes.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(isEqual):
(operator==):
(personNameComponentsForTesting):
(TEST):
(fakeDataDetectorResultForTesting):

Canonical link: https://commits.webkit.org/271585@main
